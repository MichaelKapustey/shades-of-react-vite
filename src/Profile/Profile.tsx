import { useParams, Link } from "react-router-dom"
import { useQuery } from "@tanstack/react-query";
import { graphql } from "../gql/gql";
import request from "graphql-request";
import Typography from "@mui/joy/Typography";
import Card from '@mui/joy/Card';
import ArrowBackIosNewRoundedIcon from '@mui/icons-material/ArrowBackIosNewRounded';
import LocationOnRoundedIcon from '@mui/icons-material/LocationOnRounded';
import LocationCityRoundedIcon from '@mui/icons-material/LocationCityRounded';
import DeadIcon from '@mui/icons-material/NoAccounts';
import AliveIcon from '@mui/icons-material/AccountCircleRounded';
import UnknownIcon from '@mui/icons-material/HelpRounded';
import CardCover from "@mui/joy/CardCover";
import IconButton from "@mui/joy/IconButton";
import Stack from "@mui/joy/Stack";
import { TinyCard } from "../components/TinyCard";
import { Page } from "../components/Page";
import { Img } from "react-image";

const character = graphql(/* GraphQL */ `
query getCharacter($id:ID!){
  character(id:$id){
    name,
    image,
    origin{
      name
    },
    status,
    species,
    type,
    location{
      name
    }
    episode{
      id
      name
    }
  }
}
`);

export const Profile = () => {
    const { id } = useParams();
    const { data } = useQuery({
        queryKey: ['character', id],
        queryFn: async ({ queryKey }) =>
            request(
                'https://rickandmortyapi.com/graphql',
                character,
                { id: queryKey[1] as string }
            ).catch(err => {
                console.error(err);
                return;
            }),
        placeholderData: (prev) => prev
    })
    return <Page>
        <Typography level='h1'
            variant='solid'
            color='primary'
            textAlign='center'
            startDecorator={<IconButton size="lg" color="primary" variant="solid" sx={{ ml: 2 }} component={Link} to="/">
                <ArrowBackIosNewRoundedIcon />
            </IconButton>}>
            Profile
        </Typography>
        <Stack sx={{ overflowY: 'auto' }}>
            <Stack spacing={2} direction="row" flexWrap="wrap" useFlexGap sx={{ p: 2 }}>
                <Card sx={{ height: 300, width: 300 }} variant="soft" color="primary">
                    <CardCover>
                        <Img src={data?.character?.image || ''} alt={data?.character?.name || ''} />
                    </CardCover>
                </Card>

                <Stack spacing={2} flexWrap="wrap" useFlexGap sx={{ mt: 'auto', mb: 'auto' }}>
                    <Typography level="h2">{data?.character?.name}</Typography>
                    <Typography level="title-lg">{data?.character?.type || data?.character?.species}</Typography>
                    <Stack direction="row" spacing={3} flexWrap="wrap" useFlexGap>
                        <TinyCard icon={<LocationOnRoundedIcon />}
                            title="Last known location:"
                            content={data?.character?.location?.name}
                        />
                        <TinyCard icon={<LocationCityRoundedIcon />}
                            title="Origin:"
                            content={data?.character?.origin?.name}
                        />
                        <TinyCard icon={
                            data?.character?.status === 'Alive'
                                ? <AliveIcon />
                                : data?.character?.status === 'Dead'
                                    ? <DeadIcon />
                                    : <UnknownIcon />}
                            title="Status:"
                            content={data?.character?.status}
                        />
                    </Stack>
                </Stack>
            </Stack>
            <Stack sx={{ p: 2 }}  >
                <Typography level="title-lg">Episodes:</Typography>
                <Typography>
                    {(data?.character?.episode || []).map(e => e?.name).join(', ')}
                </Typography>
            </Stack>
        </Stack>
    </Page >
}