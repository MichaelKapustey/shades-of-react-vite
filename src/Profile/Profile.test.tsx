import { render } from '@testing-library/react';
import { Profile } from './Profile';
import { vi } from 'vitest';
import React from 'react';

// The purphose of this test is to check that the correct html is generated and that all the data is in correct places
// However react-testing-library author does not support shallow renderering
// Generating huge snapshots makes them unusable. From my experience, no one reviews 1000+ lines of code, they blindly accept the change
// Thats why I have to stupidly mock every single mui dependency I've used inside Profile page

// #region Mocks

vi.mock('@tanstack/react-query', () => ({
    useQuery: () => ({
        data: {
            character: {
                name: 'The Character',
                image: 'http://some.png',
                origin: {
                    name: 'Earth'
                },
                status: 'Alive',
                species: 'Human',
                type: 'Regular Human',
                location: {
                    name: "Earth"
                },
                episode: [{
                    id: 1, name: 'First Episode'
                }, {
                    id: 2, name: 'Second Episode'
                }]
            }
        }
    })
}))

vi.mock('react-router-dom', () => ({
    useParams: () => ({ id: 4 }),
    Link: ({ children, ...rest }: React.PropsWithChildren) => <a {...rest}>{children}</a>
}))

vi.mock('@mui/joy/Button', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <button {...rest}>{children}</button>
}))

vi.mock('@mui/joy/IconButton', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <button {...rest}>{children}</button>
}))

vi.mock('@mui/joy/CardCover', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('@mui/joy/Card', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('@mui/joy/Stack', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('@mui/joy/Typography', () => ({
    default: ({ children, ...rest }: React.PropsWithChildren) => <p {...rest}>{children}</p>
}))

vi.mock('../components/TinyCard', () => ({
    TinyCard: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('../components/Page', () => ({
    Page: ({ children, ...rest }: React.PropsWithChildren) => <div {...rest}>{children}</div>
}))

vi.mock('react-image', () => ({
    Img: ({ children, ...rest }: React.PropsWithChildren) => <img {...rest}>{children}</img>
}))

vi.mock('@mui/icons-material/ArrowBackIosNewRounded', () => ({
    default: () => <svg>{'<'}</svg>
}))

vi.mock('@mui/icons-material/LocationOnRounded', () => ({
    default: () => <svg>{'L'}</svg>
}))

vi.mock('@mui/icons-material/LocationCityRounded', () => ({
    default: () => <svg>{'C'}</svg>
}))

vi.mock('@mui/icons-material/NoAccounts', () => ({
    default: () => <svg>{'D'}</svg>
}))

vi.mock('@mui/icons-material/AccountCircleRounded', () => ({
    default: () => <svg>{'A'}</svg>
}))

vi.mock('@mui/icons-material/HelpRounded', () => ({
    default: () => <svg>{'U'}</svg>
}))

//#endregion

test('Profile renders correctly and matches snapshot', () => {
    const { asFragment } = render(<Profile />);
    expect(asFragment()).toMatchSnapshot();
});

