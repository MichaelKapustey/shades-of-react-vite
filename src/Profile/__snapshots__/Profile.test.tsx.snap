// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MyComponent renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <p
      color="primary"
      level="h1"
      startdecorator="[object Object]"
      textalign="center"
      variant="solid"
    >
      Profile
    </p>
    <div
      sx="[object Object]"
    >
      <div
        direction="row"
        flexwrap="wrap"
        spacing="2"
        sx="[object Object]"
      >
        <div
          color="primary"
          sx="[object Object]"
          variant="soft"
        >
          <div>
            <img
              alt="The Character"
              src="http://some.png"
            />
          </div>
        </div>
        <div
          flexwrap="wrap"
          spacing="2"
          sx="[object Object]"
        >
          <p
            level="h2"
          >
            The Character
          </p>
          <p
            level="title-lg"
          >
            Regular Human
          </p>
          <div
            direction="row"
            flexwrap="wrap"
            spacing="3"
          >
            <div
              content="Earth"
              icon="[object Object]"
              title="Last known location:"
            />
            <div
              content="Earth"
              icon="[object Object]"
              title="Origin:"
            />
            <div
              content="Alive"
              icon="[object Object]"
              title="Status:"
            />
          </div>
        </div>
      </div>
      <div
        sx="[object Object]"
      >
        <p
          level="title-lg"
        >
          Episodes:
        </p>
        <p>
          First Episode, Second Episode
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Profile renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <p
      color="primary"
      level="h1"
      startdecorator="[object Object]"
      textalign="center"
      variant="solid"
    >
      Profile
    </p>
    <div
      sx="[object Object]"
    >
      <div
        direction="row"
        flexwrap="wrap"
        spacing="2"
        sx="[object Object]"
      >
        <div
          color="primary"
          sx="[object Object]"
          variant="soft"
        >
          <div>
            <img
              alt="The Character"
              src="http://some.png"
            />
          </div>
        </div>
        <div
          flexwrap="wrap"
          spacing="2"
          sx="[object Object]"
        >
          <p
            level="h2"
          >
            The Character
          </p>
          <p
            level="title-lg"
          >
            Regular Human
          </p>
          <div
            direction="row"
            flexwrap="wrap"
            spacing="3"
          >
            <div
              content="Earth"
              icon="[object Object]"
              title="Last known location:"
            />
            <div
              content="Earth"
              icon="[object Object]"
              title="Origin:"
            />
            <div
              content="Alive"
              icon="[object Object]"
              title="Status:"
            />
          </div>
        </div>
      </div>
      <div
        sx="[object Object]"
      >
        <p
          level="title-lg"
        >
          Episodes:
        </p>
        <p>
          First Episode, Second Episode
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
